java/ql/src/experimental/Security/CWE/CWE-400/ThreadResourceAbuse.qhelp

```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+
+<overview>
+<p><code>Thread.sleep</code> method is used to pause the execution of current thread for specified
+time. When it is used to keep several relevant tasks in synchronization and the sleep time is 
+user-controlled data, especially in the web application context, it can be abused to cause all
+of a server's threads to sleep, leading to denial of service.</p>
+</overview>
+
+<recommendation>
+<p>To guard against this attack, consider specifying an upper range of allowed sleep time or adopting
+the producer/consumer design pattern with <code>Object.wait</code> method to avoid performance 
```

Comment:
Elaborate on what a `wait` solution would look like?

Reply:
I've updated the text by adding an Oracle concurrency link and a reference to CodeQL concurrency queries. In general, <code>Object.wait()</code> can be used with <code>Object.notify()</code> for inter-thread communication to wait for results from a long processing task without waiting for a hard-coded long time. In the web context, the most straight-forward approach is not to allow pause time to be controlled by end users, which is shown in the sample.